app/models/journable/timestamps.rb

```diff
@@ -0,0 +1,103 @@
+# Copyright (C) 2012-2022 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+# In the context of the baseline-comparison feature, this module adds convenience methods
+# to the `Journable` class in order to easy access to historic queries.
+#
+# Usage:
+#
+#     # Retrieve historic state of single work package
+#     work_package = WorkPackage.find(1)
+#     historic_work_package = work_package.at_timestamp(1.year.ago)
+#     historic_work_package.id         # => 1  (same as work_package.id)
+#     historic_work_package.historic?  # => true
+#     historic_work_package.save       # raises `ActiveRecord::ReadOnlyRecord`#
+#
+#     # Filter on historic data
+#     WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1)
+#
+# See also:
+#
+# - https://github.com/opf/openproject/pull/11243
+# - https://community.openproject.org/projects/openproject/work_packages/26448
+#
+module Journable::Timestamps
+  extend ActiveSupport::Concern
+
+  class_methods do
+    # Allows to query historic data of journables.
+    #
+    # For example, to check which work packages were assigned to a
+    # specific user one year ago, run:
+    #
+    #     WorkPackage.where(assigned_to_id: 123).at_timestamp(1.year.ago)
+    #
+    def at_timestamp(timestamp)
+      Journable::HistoricActiveRecordRelation.new(all, timestamp:)
+    end
+  end
+
+  # Instantiates a journable with historic data from the given timestap.
+  #
+  #     WorkPackage.find(1).at_timestamp(1.year.ago)
+  #
+  def at_timestamp(timestamp)
+    if journal = journals.at_timestamp(timestamp).first
+      attributes = journal.data.attributes.merge(
+        {
+          "id" => id,
+          "created_at" => created_at,
+          "updated_at" => journal.updated_at,
+          "timestamp" => timestamp,
+          "position" => (position if respond_to? :position)
+        }
```

Comment:
Possibly related: https://github.com/opf/openproject/commit/e9c6bce775a3bbe53d7a2560a89232172114ee5b

With this change, the columns missing in the journal data are returned as `null`, which prevents errors, but does not provide data for these columns, yet.

@ulferts Could you explain what the `version` and the `lock_version` attributes are, and how they are related?

Do all journables have a `version` column? Then, this mapping could be added here; otherwise, I could add the modification to a subclass.

Reply:
@fiedl , the `version` column is part of the `journals` table. So every journable will have it as all are using the `journals` table. The `lock_version` on the other hand is `WorkPackage` specific. It is part of [AR::OptimisticLocking](https://api.rubyonrails.org/classes/ActiveRecord/Locking/Optimistic.html).

So regarding your question, adding a mapping here can not be generalized as only WorkPackage has the `lock_version` column. 

In the current implementation, where `null` is returned from the DB, that value is unfortunately mapped to `0` in the ruby world as that is the default value for that column. 

I don't care too much about that column's information it is just that it might lead to surprising behaviour. So if we could design some mechanism that clearly communicates to not trust that value that would be good enough for me.

Reply:
*Phone call 2023-02-02*:

- Retrieving the historic value for the `lock_version` column is not possible because this column is not journalized.
- We initialize the columns that are missing in the journalized table as `NULL` so that no database error is raised
- When trying to `pluck` or to retrieve such a value, we log a warning ➝ https://github.com/opf/openproject/pull/11243/commits/41470cac7faebde554ebe98c7a1539adffef2bc9, https://github.com/opf/openproject/pull/11243/commits/fe8ff8a1c12cf42bcebbda91d801808fb5c47036